### PR TITLE
[specific ci=6-04-Create-Basic] Add check for nil x509 leaf in vic-machine create

### LINKED ISSUE #3966: Handle failure to load x509 certificate leaf during VCH creation
**Story**
As a customer of VIC I want to be able to use custom host certificates for my VCH.

**Detail**
When specifying `--cert` and `--key` the following was observed:
```
←[34mINFO←[0m[2017-02-14T09:42:40Z] ### Installing VCH ####
time=2017-02-14T09:42:40.320460200Z level=debug msg=[BEGIN] [github.com/vmware/vic/cmd/vic-machine/create.(*Create).processParams:495]
←[34mINFO←[0m[2017-02-14T09:42:40Z] vSphere password for <snip>:
←[33mWARN←[0m[2017-02-14T09:42:47Z] Using administrative user for VCH operation - use --ops-user to improve security (see -x for advanced help)
←[37mDEBU←[0m[2017-02-14T09:42:47Z] public network: IP {"<snip>" "<snip>"} gateway {"<snip>" "<snip>"} dest: []
←[37mDEBU←[0m[2017-02-14T09:42:47Z] VCH DNS servers: [<snip>]
time=2017-02-14T09:42:47.421605800Z level=debug msg=[BEGIN] [github.com/vmware/vic/cmd/vic-machine/create.(*Create).loadCertificates:924]
time=2017-02-14T09:42:47.451629200Z level=debug msg=[BEGIN] [github.com/vmware/vic/pkg/certificate.ParseCertificate:245]
time=2017-02-14T09:42:47.453630700Z level=debug msg=[ END ] [github.com/vmware/vic/pkg/certificate.ParseCertificate:245] [2.0015ms]
time=2017-02-14T09:42:47.455638700Z level=debug msg=[ END ] [github.com/vmware/vic/cmd/vic-machine/create.(*Create).loadCertificates:924] [34.0329ms]
time=2017-02-14T09:42:47.457633900Z level=debug msg=[ END ] [github.com/vmware/vic/cmd/vic-machine/create.(*Create).processParams:495] [7.1371737s]
←[31mERRO←[0m[2017-02-14T09:42:47Z] --------------------
←[31mERRO←[0m[2017-02-14T09:42:47Z] vic-machine-windows.exe failed: runtime error: invalid memory address or nil pointer dereference
```

A speculative fix that added a check [here](https://github.com/vmware/vic/blob/2a0f9352d8501b79c510b2d26bc9856c3fba9a64/cmd/vic-machine/create/create.go#L991) for a nil Leaf got us past this issue.
It's not clear why we were unable to load the x509 leaf.

**Acceptance**
- [x] Code change to warn on failure to load x509 Leaf - should note inability to confirm certificate CommonName and --tls-cname match
- [x] VCH should be deployed successfully after warning
- [x] Test that uses certificates that result in this failure - bug1812313 will try to determine what about these certificates causes this failure.

### PR DESCRIPTION
**Pull request by andrewtchin**
Fixes #3966 
Previously if the server certificate was not the first PEM block, we would have a nil pointer deref
Check for a nil cert and warn the user, but allow install to continue. If the user is not using --no-tlsverify, they will receive an error that is being addressed in another issue and the VCH will not be usable.
Added integration test

```
⇒  bin/vic-machine-linux create -t 'root:password@192.168.218.131'  --name oceanlab   --thumbprint=C1:7A:E4:A4:22:B8:8A:68:F9:A4:E4:C2:38:91:3B:0D:8E:E0:C8:FB --no-tlsverify
Feb 21 2017 14:04:33.530-08:00 INFO  ### Installing VCH ####
Feb 21 2017 14:04:33.530-08:00 WARN  Using administrative user for VCH operation - use --ops-user to improve security (see -x for advanced help)
Feb 21 2017 14:04:33.531-08:00 WARN  Failed to load x509 leaf: Unable to confirm server certificate cname matches provided cname "". Continuing...
Feb 21 2017 14:04:33.531-08:00 INFO  Loaded server certificate oceanlab/server-cert.pem
Feb 21 2017 14:04:33.531-08:00 WARN  Configuring without TLS verify - certificate-based authentication disabled
Feb 21 2017 14:04:33.654-08:00 INFO  Validating supplied configuration
Feb 21 2017 14:04:33.740-08:00 INFO  Using default datastore: datastore1
Feb 21 2017 14:04:33.801-08:00 INFO  Firewall status: ENABLED on "/ha-datacenter/host/localhost.localdomain/localhost.localdomain"
Feb 21 2017 14:04:33.808-08:00 INFO  Firewall configuration OK on hosts:
Feb 21 2017 14:04:33.808-08:00 INFO  	"/ha-datacenter/host/localhost.localdomain/localhost.localdomain"
Feb 21 2017 14:04:33.824-08:00 WARN  Evaluation license detected. VIC may not function if evaluation expires or insufficient license is later assigned.
Feb 21 2017 14:04:33.824-08:00 INFO  License check OK
Feb 21 2017 14:04:33.824-08:00 INFO  DRS check SKIPPED - target is standalone host
Feb 21 2017 14:04:33.858-08:00 INFO  
Feb 21 2017 14:04:33.916-08:00 INFO  Creating Resource Pool "oceanlab"
Feb 21 2017 14:04:33.920-08:00 INFO  Creating VirtualSwitch
Feb 21 2017 14:04:33.954-08:00 INFO  Creating Portgroup
Feb 21 2017 14:04:33.985-08:00 INFO  Creating appliance on target
Feb 21 2017 14:04:33.993-08:00 INFO  Network role "client" is sharing NIC with "public"
Feb 21 2017 14:04:33.993-08:00 INFO  Network role "management" is sharing NIC with "public"
Feb 21 2017 14:04:34.174-08:00 INFO  Uploading images for container
Feb 21 2017 14:04:34.174-08:00 INFO  	"/home/chin/go/src/github.com/vmware/vic/bin/bootstrap.iso"
Feb 21 2017 14:04:34.174-08:00 INFO  	"/home/chin/go/src/github.com/vmware/vic/bin/appliance.iso"
Feb 21 2017 14:04:41.773-08:00 INFO  Waiting for IP information
Feb 21 2017 14:04:54.164-08:00 INFO  Waiting for major appliance components to launch
Feb 21 2017 14:04:54.356-08:00 INFO  Checking VCH connectivity with vSphere target
Feb 21 2017 14:04:54.796-08:00 INFO  vSphere API Test: https://192.168.218.131 vSphere API target responds as expected
Feb 21 2017 14:04:54.796-08:00 INFO  Checking Docker API endpoint
Feb 21 2017 14:04:57.854-08:00 INFO  Initialization of appliance successful
Feb 21 2017 14:04:57.854-08:00 INFO  
Feb 21 2017 14:04:57.854-08:00 INFO  VCH Admin Portal:
Feb 21 2017 14:04:57.854-08:00 INFO  https://192.168.218.133:2378
Feb 21 2017 14:04:57.854-08:00 INFO  
Feb 21 2017 14:04:57.854-08:00 INFO  Published ports can be reached at:
Feb 21 2017 14:04:57.854-08:00 INFO  192.168.218.133
Feb 21 2017 14:04:57.854-08:00 INFO  
Feb 21 2017 14:04:57.854-08:00 INFO  Docker environment variables:
Feb 21 2017 14:04:57.854-08:00 INFO  DOCKER_HOST=192.168.218.133:2376
Feb 21 2017 14:04:57.854-08:00 INFO  
Feb 21 2017 14:04:57.854-08:00 INFO  Environment saved in oceanlab/oceanlab.env
Feb 21 2017 14:04:57.854-08:00 INFO  
Feb 21 2017 14:04:57.854-08:00 INFO  Connect to docker:
Feb 21 2017 14:04:57.854-08:00 INFO  docker -H 192.168.218.133:2376 --tls info
Feb 21 2017 14:04:57.854-08:00 INFO  Installer completed successfully
```

an example server-cert.pem that would cause this is:
```
-----BEGIN RSA PRIVATE KEY-----
JUNK
-----END RSA PRIVATE KEY-----
-----BEGIN CERTIFICATE-----
MIIDAjCCAeqgAwIBAgIQapIxSHwtGxNH0M9BaAh+kzANBgkqhkiG9w0BAQsFADAS
MRAwDgYDVQQKEwdkZWZhdWx0MB4XDTE3MDIyMDIyMDA0M1oXDTE4MDIyMTIyMDA0
M1owEjEQMA4GA1UEChMHZGVmYXVsdDCCASIwDQYJKoZIhvcNAQEBBQADggEPADCC
AQoCggEBAObOltkeRNa3UhxxDArqctWkv1DOcx6zdSKtFc92b6QjQwVosm2NA4YN
Mk+3yLjiVTDM7J49I8vxOpdAA/Yx58Qnc79cqjZVIMSg0nMreCWemPM0IW4iYO1K
ngVBX6iP6RsYMg9xiyEesKg+RtLZXjaz7O0lEwvYJjLrUj5Vo/0xiWdOJ4YC+F5d
Wr4iwbbTjfqj120opvd+GaA/UdcGgGBOIxPKdF7xSdYoY/fwgNtfOKxOa9JBCU17
WALpdKeYFv74awl1jR7ZgKwYmAEEGTr3AGAqrUeD0/MlN1/Zrxx1DmabtkAkrCol
fWKndk6waBs8t9/V3AA4GP/d9HvU9ScCAwEAAaNUMFIwDgYDVR0PAQH/BAQDAgOo
MBMGA1UdJQQMMAoGCCsGAQUFBwMBMAwGA1UdEwEB/wQCMAAwHQYDVR0OBBYEFKOs
ykvZgVngSY6xFcU4rR8jyBr4MA0GCSqGSIb3DQEBCwUAA4IBAQADnLeT4bPwuhYA
lLVDWdc2VWHRByJtQnvccWK97JoH9mdP5oUmTtK5SJSbmA4/FJbmXn3U5LGc4nPM
SAiqDWqgpfcI7Hwj/C9EsDKkmmaNmsfBw1D9sTRR99yCsR/3PD1vmyi2R35sqjO+
evmipP1/y+nReo95yauluai3LxgIKsIYPlDUc+VezNTNvSx9kNFtGI+jF0og0eI+
gRH9YnU7YahE78WJA6ymE95R/P3GD2QdQE1Ap5FMWFsi/Yb1qtM1O0SLiBCJii9W
fh3pB3RiqMCq8QHjy9smkJD1CkO6vxOsI0r+k7CiG2HUUVNS1aSA67QI69E/FkUe
GwUQ1Emg
-----END CERTIFICATE-----
```

### CODE REVIEW

File: cmd/vic-machine/create/create.go [hunk 1]
```diff
@@ -986,14 +986,18 @@ func (c *Create) loadCertificates() ([]byte, *certificate.KeyPair, error) {
...
 	}
 
-	// We just do a direct equality check here - trying to be clever is liable to lead to hard
-	// to diagnose errors
-	if cert.Leaf.Subject.CommonName != c.cname {
-		log.Errorf("Provided cname does not match that in existing server certificate: %s", cert.Leaf.Subject.CommonName)
-		if c.Debug.Debug > 2 {
-			log.Debugf("Certificate does not match provided cname: %#+v", cert.Leaf)
+	if cert.Leaf == nil {
+		log.Warnf("Failed to load x509 leaf: Unable to confirm server certificate cname matches provided cname %q. Continuing...", c.cname)
```
Comment (emlin) [inline cmd/vic-machine/create/create.go]: If the Leaf is empty, can we do other checks here?
If the common name does not match, we'll return failure, but if we cannot verify, just continue. Feels a bit wrong here.

Comment (hickeng) [inline cmd/vic-machine/create/create.go]: If the common name doesn't match then it's an actual error - if we cannot determine whether they match then we cannot say for sure that it's an error. The point of greater concern is why we cannot parse the certificate.

File: cmd/vic-machine/create/create.go [hunk 2]
```diff
@@ -986,14 +986,18 @@ func (c *Create) loadCertificates() ([]byte, *certificate.KeyPair, error) {
 		return certs, nil, err
 	}
```
Comment (anchal-agrawal) [inline cmd/vic-machine/create/create.go]: Sidenote: we should update this file's copyright year.

Comment (andrewtchin) [inline cmd/vic-machine/create/create.go]: fixed

File: cmd/vic-machine/create/create.go (hunk 1 above)
Comment (andrewtchin) [inline cmd/vic-machine/create/create.go]: I tried adding a debug log in keypair.Certificate() on L983, but that doesn't seem to provide any useful information

**Comment by emlin** [comment]
The code change is fine to fix nil pointer issue, but it does not actually solve the certificate validation issue
I assume there are some mismatch between our validation and customer certificate, but if we do not have details there, it's hard to fix it. So I'm fine to have this to avoid panic at least.

**Pull request merged**